Rust/src/random_cut_forest.rs

```diff
@@ -0,0 +1,250 @@
+/// streams. This type is the main interface for training a random cut forest
+/// and using the model for various scoring operations such as anomaly detection.
+///
+/// A random cut forest is a collection of random cut trees; specifically a
+/// collection of [`SampledTree`] structs. An update to a random cut forest
+/// model corresponds to independently updating each `SampledTree` with the
+/// input point. When a scoring algorithm is called, such as anomaly score,
+/// each random cut tree reports a score and these scores are aggegated in
+/// some way particular to the scoring algorithm.
+///
+/// It is recommended to use [`RandomCutForestBuilder`] to create a new
+/// [`RandomCutForest`] model.
+///
+/// # Examples
+///
+/// ```
+/// use random_cut_forest::{RandomCutForest, RandomCutForestBuilder};
+///
+/// // create the default random cut forest on three-dimensional data points
+/// let mut forest: RandomCutForest<f32> = RandomCutForestBuilder::new(3).build();
+///
+/// // update the forest with some data points
+/// forest.update(vec![0.0, 0.0, 0.0]);
+/// forest.update(vec![1.0, 1.0, 1.0]);
+/// forest.update(vec![0.0, 0.5, 0.3]);
+/// forest.update(vec![0.6, 0.0, -0.2]);
+///
+/// // compute anomaly scores
+/// // let score = forest.anomaly_score(vec![0.1, 0.2, 0.3]);
+/// ```
+pub struct RandomCutForest<T> {
+    dimension: usize,
+    num_observations: usize,
+    num_trees: usize,
+    sample_size: usize,
+    time_decay: f32,
+    trees: Vec<SampledTree<T>>,
+}
+
+impl<T> RandomCutForest<T>
+    where T: Float + Sum
+{
+
+    /// Update a random cut forest with a new data point.
+    ///
+    /// A copy of the data point will be sent to each sampled tree in the forest
+    /// for consideration. Each tree independently decides whether to accept the
+    /// point into its sample.
+    ///
+    /// # Panics
+    ///
+    /// If the dimensionality of the input data point does not match the
+    /// dimensionality of the forest.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use random_cut_forest::{RandomCutForest, RandomCutForestBuilder};
+    ///
+    /// // create a default RCF on two-dimensional data
+    /// let mut forest: RandomCutForest<f32> = RandomCutForestBuilder::new(2).build();
+    ///
+    /// // update the forest with some data
+    /// forest.update(vec![0.0, 0.0]);
+    /// forest.update(vec![1.0, 0.0]);
+    /// forest.update(vec![0.0, 1.0]);
+    /// forest.update(vec![1.0, 1.0]);
+    ///
+    /// // update panics if the point has an incorrect dimensionality/length
+    /// //forest.update(vec![2.0, 3.0, 4.0]);
+    /// ```
+    pub fn update(&mut self, point: Vec<T>) {
+        assert_eq!(point.len(), self.dimension,
+            "Dimension mismatch. Expected {}-dimensional input.",
+            self.dimension);
+
+        self.num_observations += 1;
+        for tree in self.trees.iter_mut() {
+            tree.update(point.clone(), self.num_observations)
```

Comment:
If each tree has the same number of observations, does the tree need to store this information or can it just be stored in the rcf? I believe the sampler will store num_observations as well.

Reply:
I believe these trees are "sampled trees", meaning this is the operation submitting the point to the sampler. We want the sampler to receive (point, index) pairs for data debugging, and we want to take the index as a parameter (instead of using a counter in the sampler) for flexibility. In the future we may want to support out-of-order updates (e.g., when processing data in batch mode).

Reply:
Makes sense, thanks